[ASDisplayNode] Allow measure always be off the main thread

maicki opened the pull request:
Currently measurement always needs to happen on the main thread if implicit hierarchy management is enabled as adding and removing from nodes needs to happen on the main thread. We now will trampoline to the main thread to do the insertion and deletion of nodes.

This also resolves the issue that can occur if a node is already loaded deep in the layout hierarchy in the layout that the node is transforming to. Before insertion or deletion is happening we need to crawl the layout hierarchy to check that though.

*Open things:*
* This theoretically that would enable us always measure nodes in the data controller asynchronous
* Furthermore this also would us allow to get rid of the `shouldMeasureAsync` parameter in `transitionLayoutWithAnimation:shouldMeasureAsync:measurementCompletion:` that I think should not be exposed at all - @levi any thoughts on that?
* Do we need to check somehow the transition id in `_applyLayout:layoutTransition:` to abort the insertion or deletion of nodes if the block runs on the main thread?
* Would it be worth to have it's own transition queue instead of always using one of the global background queues?

@appleguy @Adlai-Holler @levi Would be great if I could get a review and thoughts

Adlai-Holler commented:
If applyLayout:layoutTransition is called on a background thread, is it possible that a subnode involved in the transition could get its node loaded on the main thread after `ASLayoutCanTransitionAsynchronous` but before `applySubnodeTransition`?

levi commented on AsyncDisplayKit/ASDisplayNode.mm:
```diff
@@ -838,8 +846,8 @@ - (void)setUsesImplicitHierarchyManagement:(BOOL)value
 
 - (BOOL)_hasTransitionInProgress
```
This should be _isTransitionInProgress

levi commented on AsyncDisplayKit/Private/ASLayoutTransition.h:
```diff
@@ -18,16 +18,45 @@
+/**
+ * Returns if the layout transition can happen asynchronously
+ */
+@property (nonatomic, readonly, assign) BOOL canTransitionAsynchronous;
```
`@property (nonatomic, readonly, assign) BOOL isSynchronous;`

levi commented on AsyncDisplayKit/Private/ASLayoutTransition.mm:
```diff
@@ -18,10 +18,37 @@
+ * Search the whole layout stack if at least one layout has a layoutable object that can not be layed out asynchronous.
+ * This can be the case for example if a node was already loaded
+ */
+static BOOL ASLayoutCanTransitionAsynchronous(ASLayout *layout) {
```
Make inline

levi commented on AsyncDisplayKit/Private/ASLayoutTransition.h:
```diff
@@ -18,16 +18,45 @@
+/**
+ * Insert and remove subnodes that where added or removed between the previousLayout and the pendingLayout
+ */
+- (void)applySubnodeTransition;
```
`completeTransition` or something similar. Would be great to add a level of indirection in the naming such that the concept of subnode management isn't exposed to the caller.

maicki replied:
I named it `startTransition` now. We should also look into finding a way to not have to expose `applySubnodeInsertions` and `applySubnodeRemovals` to the consumer.

maicki commented:
@Adlai-Holler Theoretically this could happen, in fact this can happen with the current implementation too. So in this nothing will change. I don't know at the moment a good way to fix that, but we definitely should consider it in the broader refactoring around layout transitions that I'm working on.

@levi Addressed your comments

@levi @appleguy @Adlai-Holler Can I get another look on it. Thanks!

Adlai-Holler commented:
Awesome! Time to 🚢
